# Assignment 4 Extra credit

Krishan Subudhi (ksubudhi@uw.edu)

Student number: 2040900

---

## 7. predicted RNAs that are not annotated in the GFF file

Another annotation source:

Along with NCBI I found another source for the genome which was updated recently. 

ftp://ftp.ensemblgenomes.org/pub/bacteria/release-48/gff3/bacteria_0_collection/methanocaldococcus_jannaschii_dsm_2661

Methanocaldococcus_jannaschii_dsm_2661.ASM9166v1.46.gff3.gz

Source : http://bacteria.ensembl.org/Methanocaldococcus_jannaschii_dsm_2661/Info/Index


</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>state</th>
      <th>start</th>
      <th>end</th>
      <th>length</th>
      <th>percentage_overlap</th>
      <th>percentage_overlap2</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>State 2</td>
      <td>97326</td>
      <td>97541</td>
      <td>216</td>
      <td>51.851852</td>
      <td>51.851852</td>
    </tr>
    <tr>
      <th>1</th>
      <td>State 2</td>
      <td>97627</td>
      <td>97823</td>
      <td>197</td>
      <td>44.670051</td>
      <td>44.670051</td>
    </tr>
    <tr>
      <th>2</th>
      <td>State 2</td>
      <td>111764</td>
      <td>111856</td>
      <td>93</td>
      <td>95.698925</td>
      <td>95.698925</td>
    </tr>
    <tr>
      <th>3</th>
      <td>State 2</td>
      <td>118079</td>
      <td>118179</td>
      <td>101</td>
      <td>0.000000</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>4</th>
      <td>State 2</td>
      <td>138345</td>
      <td>138419</td>
      <td>75</td>
      <td>100.000000</td>
      <td>100.000000</td>
    </tr>
    <tr>
      <th>5</th>
      <td>State 2</td>
      <td>154610</td>
      <td>157697</td>
      <td>3088</td>
      <td>96.437824</td>
      <td>96.437824</td>
    </tr>
    <tr>
      <th>6</th>
      <td>State 2</td>
      <td>157782</td>
      <td>159591</td>
      <td>1810</td>
      <td>85.801105</td>
      <td>85.801105</td>
    </tr>
    <tr>
      <th>7</th>
      <td>State 2</td>
      <td>186974</td>
      <td>187067</td>
      <td>94</td>
      <td>94.680851</td>
      <td>94.680851</td>
    </tr>
    <tr>
      <th>8</th>
      <td>State 2</td>
      <td>190831</td>
      <td>190907</td>
      <td>77</td>
      <td>98.701299</td>
      <td>98.701299</td>
    </tr>
    <tr>
      <th>9</th>
      <td>State 2</td>
      <td>215200</td>
      <td>215296</td>
      <td>97</td>
      <td>89.690722</td>
      <td>89.690722</td>
    </tr>
    <tr>
      <th>10</th>
      <td>State 2</td>
      <td>227705</td>
      <td>227782</td>
      <td>78</td>
      <td>97.435897</td>
      <td>97.435897</td>
    </tr>
    <tr>
      <th>11</th>
      <td>State 2</td>
      <td>291972</td>
      <td>291997</td>
      <td>26</td>
      <td>0.000000</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>12</th>
      <td>State 2</td>
      <td>303990</td>
      <td>304080</td>
      <td>91</td>
      <td>97.802198</td>
      <td>97.802198</td>
    </tr>
    <tr>
      <th>13</th>
      <td>State 2</td>
      <td>358766</td>
      <td>358942</td>
      <td>177</td>
      <td>85.875706</td>
      <td>85.875706</td>
    </tr>
    <tr>
      <th>14</th>
      <td>State 2</td>
      <td>359974</td>
      <td>360046</td>
      <td>73</td>
      <td>100.000000</td>
      <td>100.000000</td>
    </tr>
    <tr>
      <th>15</th>
      <td>State 2</td>
      <td>402969</td>
      <td>403057</td>
      <td>89</td>
      <td>85.393258</td>
      <td>85.393258</td>
    </tr>
    <tr>
      <th>16</th>
      <td>State 2</td>
      <td>412582</td>
      <td>412635</td>
      <td>54</td>
      <td>0.000000</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>17</th>
      <td>State 2</td>
      <td>552537</td>
      <td>552862</td>
      <td>326</td>
      <td>96.932515</td>
      <td>96.932515</td>
    </tr>
    <tr>
      <th>18</th>
      <td>State 2</td>
      <td>619161</td>
      <td>619236</td>
      <td>76</td>
      <td>97.368421</td>
      <td>97.368421</td>
    </tr>
    <tr>
      <th>19</th>
      <td>State 2</td>
      <td>637579</td>
      <td>638153</td>
      <td>575</td>
      <td>81.043478</td>
      <td>81.043478</td>
    </tr>
    <tr>
      <th>20</th>
      <td>State 2</td>
      <td>638334</td>
      <td>640132</td>
      <td>1799</td>
      <td>86.492496</td>
      <td>86.492496</td>
    </tr>
    <tr>
      <th>21</th>
      <td>State 2</td>
      <td>640217</td>
      <td>643449</td>
      <td>3233</td>
      <td>95.731519</td>
      <td>95.731519</td>
    </tr>
    <tr>
      <th>22</th>
      <td>State 2</td>
      <td>643500</td>
      <td>643767</td>
      <td>268</td>
      <td>96.268657</td>
      <td>96.268657</td>
    </tr>
    <tr>
      <th>23</th>
      <td>State 2</td>
      <td>763767</td>
      <td>763845</td>
      <td>79</td>
      <td>96.202532</td>
      <td>96.202532</td>
    </tr>
    <tr>
      <th>24</th>
      <td>State 2</td>
      <td>764022</td>
      <td>764095</td>
      <td>74</td>
      <td>100.000000</td>
      <td>100.000000</td>
    </tr>
    <tr>
      <th>25</th>
      <td>State 2</td>
      <td>774708</td>
      <td>774788</td>
      <td>81</td>
      <td>0.000000</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>26</th>
      <td>State 2</td>
      <td>863476</td>
      <td>864151</td>
      <td>676</td>
      <td>73.816568</td>
      <td>73.816568</td>
    </tr>
    <tr>
      <th>27</th>
      <td>State 2</td>
      <td>873579</td>
      <td>873778</td>
      <td>200</td>
      <td>39.000000</td>
      <td>39.000000</td>
    </tr>
    <tr>
      <th>28</th>
      <td>State 2</td>
      <td>883675</td>
      <td>883755</td>
      <td>81</td>
      <td>91.358025</td>
      <td>91.358025</td>
    </tr>
    <tr>
      <th>29</th>
      <td>State 2</td>
      <td>951852</td>
      <td>951968</td>
      <td>117</td>
      <td>0.000000</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>30</th>
      <td>State 2</td>
      <td>1038544</td>
      <td>1038622</td>
      <td>79</td>
      <td>97.468354</td>
      <td>97.468354</td>
    </tr>
    <tr>
      <th>31</th>
      <td>State 2</td>
      <td>1129124</td>
      <td>1129194</td>
      <td>71</td>
      <td>0.000000</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>32</th>
      <td>State 2</td>
      <td>1150142</td>
      <td>1150402</td>
      <td>261</td>
      <td>58.237548</td>
      <td>58.237548</td>
    </tr>
    <tr>
      <th>33</th>
      <td>State 2</td>
      <td>1189943</td>
      <td>1190054</td>
      <td>112</td>
      <td>98.214286</td>
      <td>98.214286</td>
    </tr>
    <tr>
      <th>34</th>
      <td>State 2</td>
      <td>1313165</td>
      <td>1313251</td>
      <td>87</td>
      <td>97.701149</td>
      <td>97.701149</td>
    </tr>
    <tr>
      <th>35</th>
      <td>State 2</td>
      <td>1659451</td>
      <td>1659520</td>
      <td>70</td>
      <td>0.000000</td>
      <td>0.000000</td>
    </tr>
  </tbody>
</table>
</div>



It looks like both the sources (NCBI, bacteria.ensemble.org ) point to the same annotation

1. [RFAM](http://rfam.xfam.org/search?q=UP000000805%20AND%20entry_type:%22Sequence%22%20AND%20rna_type:%22tRNA%22) showed the tRNAs (generally very small sequences) but there was no way to download them. From a brief overview it looks like, the website does not have much information. 
2. [RNACentral](https://rnacentral.org/export/results?job=a1136a5f-3857-42ef-a88b-2b78bd024bd7) lets you download all the searches but the downloaded file does not have position information in it.



```python
df[df.percentage_overlap <50]
```




<div>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>state</th>
      <th>start</th>
      <th>end</th>
      <th>length</th>
      <th>percentage_overlap</th>
      <th>percentage_overlap2</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>1</th>
      <td>State 2</td>
      <td>97627</td>
      <td>97823</td>
      <td>197</td>
      <td>44.670051</td>
      <td>44.670051</td>
    </tr>
    <tr>
      <th>3</th>
      <td>State 2</td>
      <td>118079</td>
      <td>118179</td>
      <td>101</td>
      <td>0.000000</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>11</th>
      <td>State 2</td>
      <td>291972</td>
      <td>291997</td>
      <td>26</td>
      <td>0.000000</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>16</th>
      <td>State 2</td>
      <td>412582</td>
      <td>412635</td>
      <td>54</td>
      <td>0.000000</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>25</th>
      <td>State 2</td>
      <td>774708</td>
      <td>774788</td>
      <td>81</td>
      <td>0.000000</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>27</th>
      <td>State 2</td>
      <td>873579</td>
      <td>873778</td>
      <td>200</td>
      <td>39.000000</td>
      <td>39.000000</td>
    </tr>
    <tr>
      <th>29</th>
      <td>State 2</td>
      <td>951852</td>
      <td>951968</td>
      <td>117</td>
      <td>0.000000</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>31</th>
      <td>State 2</td>
      <td>1129124</td>
      <td>1129194</td>
      <td>71</td>
      <td>0.000000</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>35</th>
      <td>State 2</td>
      <td>1659451</td>
      <td>1659520</td>
      <td>70</td>
      <td>0.000000</td>
      <td>0.000000</td>
    </tr>
  </tbody>
</table>
</div>



Except for 27, none of the predictions are very long. So the model is not missing major genes.  

## 8. Evaluation Metric
Q. Do a more formal analysis of the "accuracy" of this method for its designated purpose of discovering RNA genes. I.e., count True Positives, False Negatives, etc. 

Based on the paper, *Assessing computational tools for the discovery of transcription factor binding sites* I have calculate **Specificity and PPV metrics** at neucleotite levels. 

* nTP is the number of nucleotide positions in both known sites and
predicted sites,
* nFN is the number of nucleotide positions in known sites but not in
predicted sites,
* nFP is the number of nucleotide positions not in known sites but in
predicted sites, and
* nTN is the number of nucleotide positions in neither known sites
nor predicted sites.


```python
df['overlap_length'] = overlap_lengths
df.head()
```




<div>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>state</th>
      <th>start</th>
      <th>end</th>
      <th>length</th>
      <th>percentage_overlap</th>
      <th>percentage_overlap2</th>
      <th>overlap_length</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>State 2</td>
      <td>97326</td>
      <td>97541</td>
      <td>216</td>
      <td>51.851852</td>
      <td>51.851852</td>
      <td>112</td>
    </tr>
    <tr>
      <th>1</th>
      <td>State 2</td>
      <td>97627</td>
      <td>97823</td>
      <td>197</td>
      <td>44.670051</td>
      <td>44.670051</td>
      <td>88</td>
    </tr>
    <tr>
      <th>2</th>
      <td>State 2</td>
      <td>111764</td>
      <td>111856</td>
      <td>93</td>
      <td>95.698925</td>
      <td>95.698925</td>
      <td>89</td>
    </tr>
    <tr>
      <th>3</th>
      <td>State 2</td>
      <td>118079</td>
      <td>118179</td>
      <td>101</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0</td>
    </tr>
    <tr>
      <th>4</th>
      <td>State 2</td>
      <td>138345</td>
      <td>138419</td>
      <td>75</td>
      <td>100.000000</td>
      <td>100.000000</td>
      <td>75</td>
    </tr>
  </tbody>
</table>
</div>




```python
genepos2.head()
```




<div>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>start</th>
      <th>end</th>
      <th>length</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>97426</td>
      <td>97537</td>
      <td>111</td>
    </tr>
    <tr>
      <th>1</th>
      <td>97629</td>
      <td>97716</td>
      <td>87</td>
    </tr>
    <tr>
      <th>2</th>
      <td>111766</td>
      <td>111854</td>
      <td>88</td>
    </tr>
    <tr>
      <th>3</th>
      <td>138344</td>
      <td>138419</td>
      <td>75</td>
    </tr>
    <tr>
      <th>4</th>
      <td>154662</td>
      <td>157639</td>
      <td>2977</td>
    </tr>
  </tbody>
</table>
</div>




```python
nTP = df.overlap_length.sum()
nFN = genepos2.length.sum() - nTP
nFP = df.length.sum() - nTP
nTN = 1664970 - (nTP+nFP+nFN)

nTP, nFN, nFP, nTN
```




    (12603, 42, 2098, 1650227)



### Sensitivity
Sensitivity gives fraction of known site neucleotites that are predicted

Sensitivity: nSn = nTP/(nTP + nFN), and


```python
nSn = nTP/(nTP + nFN)
print('Sensitivity  = ', nSn)
```

    Sensitivity  =  0.9966785290628707
    

### PPV
PPV gives fraction of predicted site neucleotites that are known

Positive Predictive Value: nPPV = nTP/(nTP + nFP)


```python
nPPV =  nTP/(nTP + nFP)
print('PPV  = ', nPPV)
```

    PPV  =  0.8572886198217808
    

Many predictions are either starting before the gene starts or still continue after the gene ends. Because of to many false positives, the PPV value is low for this model.

#### Note on PPV and Sensitivity:

For datasets with no binding sites (TP + FN = 0) , sensitivity is not defined while PPV is uninformative. 

If the model predicts no hit state ( TP + FP = 0) , PPV is undefined and sensitivity is undefined.
